test/functional/lib/ipfs-proxy/access-control.test.js

```diff
@@ -0,0 +1,134 @@
+'use strict'
+const { describe, it } = require('mocha')
+const { expect } = require('chai')
+const AccessControl = require('../../../../add-on/src/lib/ipfs-proxy/access-control')
+const Storage = require('mem-storage-area/Storage')
+
+describe('lib/ipfs-proxy/access-control', () => {
+  it('should maintain ACL', async () => {
+    const accessControl = new AccessControl(new Storage())
+    let acl = await accessControl.getAcl()
+
+    // Ensure ACL starts out empty
+    expect(acl).to.deep.equal({})
+
+    const sets = [
+      ['http://ipfs.io', 'ipfs.files.add', true],
```

Comment:
Worth ensuring that urls are normalised correctly...

are

- `http://ipfs.io
- `http://ipfs.io/`
- `http://ipfs.io/ipfs/QmHash`
- `https://ipfs.io/`

equivalent for permissions purposes? are

- `https://www.ipfs.io`
- `https://ipfs.io`

different for permissions purposes?

Reply:
The origin for permissions purposes is [extracted out of the URL that is reported by the port](https://github.com/tableflip/ipfs-companion/blob/af594639a81a87e54f5de1294133f11d29750c9b/add-on/src/lib/ipfs-proxy/index.js#L41) so this information doesn't come from "outside", and nor does the permission name (comes from [`ACL_FUNCTIONS`](https://github.com/tableflip/ipfs-companion/blob/af594639a81a87e54f5de1294133f11d29750c9b/add-on/src/lib/ipfs-proxy/index.js#L48-L51)) or allow/deny decision (comes from [`window.confirm`](https://github.com/tableflip/ipfs-companion/blob/af594639a81a87e54f5de1294133f11d29750c9b/add-on/src/lib/ipfs-proxy/index.js#L130)).

So the origin _is_ normalised already by virtue of parsing the URL and extracting the origin. For permissions we're using the same-origin policy, so URLs for a different origin need different approval.

> An origin is defined as a combination of URI scheme, host name, and port number
> https://en.wikipedia.org/wiki/Same-origin_policy

* `http://ipfs.io` and `http://ipfs.io/` are normalised to `http://ipfs.io`
* `http://ipfs.io/ipfs/QmHash` not supported as a separate "origin" yet (mentioned in "What's still todo (in a future PR)?" in the description) so origin would be `http://ipfs.io`
* `https://ipfs.io` **is a different origin** for permissions (because scheme differs)
* `https://www.ipfs.io` and `https://ipfs.io`, **also a different origin** for permissions (because host differs)

Reply:
ah sure, yes, I was trying to say that there is an implicit dependency for this to function correctly on the url having been normalised before it gets passed on to the ACL, but there is no test to assert it. 

With the new changes the ACL will error out if the user tries to set a value that isn't a valid origin as normalised by `URL`, so i guess that covers it.

Reply:
✨ 🚀 ✨